cdap-ui/app/directives/widget-container/widget-output-schema/widget-output-schema-ctrl.js

```diff
@@ -15,32 +15,50 @@
 */
 angular.module(PKG.name + '.commons')
   .controller('MyOutputSchemaCtrl', function($scope, GLOBALS, HydratorPlusPlusNodeService) {
-    if (typeof $scope.node.outputSchema === 'string') {
-      $scope.node.outputSchema = [HydratorPlusPlusNodeService.getOutputSchemaObj($scope.node.outputSchema)];
-    }
-    this.outputSchemas = $scope.node.outputSchema
-      .map((node) => {
-        var schema = node.schema;
-        if (typeof schema === 'string') {
-          try {
-            schema = JSON.parse(schema);
-          } catch(e) {
-            schema = {
-              'name': GLOBALS.defaultSchemaName,
-              'type': 'record',
-              'fields': []
+    this.formatOutputSchema = () => {
+      if (!$scope.schemaAdvance) {
+        if (typeof $scope.node.outputSchema === 'string') {
+          $scope.node.outputSchema = [HydratorPlusPlusNodeService.getOutputSchemaObj($scope.node.outputSchema)];
+        }
+        this.outputSchemas = $scope.node.outputSchema
+          .map((node) => {
+            var schema = node.schema;
+            if (typeof schema === 'string') {
+              try {
+                schema = JSON.parse(schema);
+              } catch(e) {
+                schema = {
+                  'name': GLOBALS.defaultSchemaName,
+                  'type': 'record',
+                  'fields': []
+                };
+              }
+            }
+            return {
+              name: node.name,
+              schema: schema
             };
+          });
+      } else {
+        if ($scope.node.outputSchema.length > 0 && $scope.node.outputSchema[0].schema) {
+          let schema = $scope.node.outputSchema[0].schema;
+          if (typeof schema !== 'string') {
+            schema = JSON.stringify(schema);
           }
+          this.outputSchemaString = schema;
+        } else {
+          this.outputSchemaString = '';
         }
-        return {
-          name: node.name,
-          schema: schema
-        };
-      });
+      }
+    };
+
+    this.formatOutputSchema();
 
-    this.currentIndex = 0;
+    $scope.$watch('schemaAdvance', () => {
+      this.formatOutputSchema();
+    });
 
-    this.onOutputSchemaChange = (newOutputSchemas) => {
-      $scope.node.outputSchema = newOutputSchemas;
+    this.onOutputSchemaChange = (newValue) => {
+      $scope.node.outputSchema = newValue;
```

Comment:
will the `newValue` be a string or an object? In line 22 we treat it as an array but here `ng-change` calls this function with a string.

Reply:
Depends on which component is calling it, but even if it's a string it still gets converted to an array in other stages of the code.